src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -3432,6 +3476,788 @@ public void Inheritance_17()
+                // (4,21): error CS8866: Record member 'C.Y' must be a readable instance property of type 'int' to match positional parameter 'Y'.
+                // record C(int X, int Y)
+                Diagnostic(ErrorCode.ERR_BadRecordMemberForPositionalParameter, "Y").WithArguments("C.Y", "int", "Y").WithLocation(4, 21));
+
+            Assert.Equal(
+                "void C.Deconstruct(out System.Int32 X, out System.Int32 Y)",
+                comp.GetMember("C.Deconstruct").ToTestDisplayString(includeNonNullable: false));
+        }
+        
+        [Fact]
+        public void Deconstruct_Conversion_02()
+        {
+            var source = @"
+#nullable enable
+using System;
+
+record C(string? X, string Y)
+{
+    public string X { get; init; } = null!;
+    public string? Y { get; init; }
+
+    static void M(C c)
+    {
+        switch (c)
+        {
+            case C(var x, string y):
+                Console.Write(x);
+                Console.Write(y);
+                break;
+        }
+    }
+
+    static void Main()
+    {
+        M(new C(""a"", ""b""));
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+
+            Assert.Equal(
+                "void C.Deconstruct(out System.String? X, out System.String Y)",
+                comp.GetMember("C.Deconstruct").ToTestDisplayString(includeNonNullable: false));
+        }
+
+        [Fact]
+        public void Deconstruct_Conversion_03()
+        {
+            var source = @"
+using System;
+
+class Base { }
+class Derived : Base { }
+
+record C(Derived X, Base Y)
+{
+    public Base X { get; init; }
+    public Derived Y { get; init; }
+
+    static void M(C c)
+    {
+        switch (c)
+        {
+            case C(Derived x, Base y):
+                Console.Write(x);
+                Console.Write(y);
+                break;
+        }
+    }
+
+    static void Main()
+    {
+        M(new C(new Derived(), new Base()));
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (7,18): error CS8866: Record member 'C.X' must be a readable instance property of type 'Derived' to match positional parameter 'X'.
+                // record C(Derived X, Base Y)
+                Diagnostic(ErrorCode.ERR_BadRecordMemberForPositionalParameter, "X").WithArguments("C.X", "Derived", "X").WithLocation(7, 18),
+                // (7,26): error CS8866: Record member 'C.Y' must be a readable instance property of type 'Base' to match positional parameter 'Y'.
+                // record C(Derived X, Base Y)
+                Diagnostic(ErrorCode.ERR_BadRecordMemberForPositionalParameter, "Y").WithArguments("C.Y", "Base", "Y").WithLocation(7, 26));
+
+            Assert.Equal(
+                "void C.Deconstruct(out Derived X, out Base Y)",
+                comp.GetMember("C.Deconstruct").ToTestDisplayString(includeNonNullable: false));
+        }
```

Comment:
Please add a test with a generic type where we deconstruct instances of derived types. Perhaps record types such as:
```C#
record A<T>(T P);
record B1(int P, object Q) : A<int>;
record B2(object P, object Q) : A<object>;
record B3<T>(T P, object Q) : A<T>;
```

Reply:
Done